src/lib/support/Pool.h

```diff
@@ -0,0 +1,130 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *
+ * @brief
+ *   Defines a header only BitMapObjectPool.
+ */
+
+#pragma once
+
+#include <array>
+#include <assert.h>
+#include <atomic>
+#include <new>
+#include <stddef.h>
+
+namespace chip {
+
+template <class T, size_t N>
+class BitMapObjectPool
+{
+public:
+    BitMapObjectPool()
+    {
+        for (size_t word = 0; word * kAtomicSize < N; ++word)
+        {
+            mUsage[word].store(0);
+        }
+    }
+
+    static size_t Size() { return N; }
+
+    template <typename... Args>
+    T * New(Args &&... args)
+    {
+        for (size_t word = 0; word * kAtomicSize < N; ++word)
+        {
+            auto & usage   = mUsage[word];
+            uint32_t value = usage.load();
+            for (size_t offset = 0; offset < kAtomicSize && offset + word * kAtomicSize < N; ++offset)
+            {
+                if ((value & (1 << offset)) == 0)
+                {
+                    if (usage.compare_exchange_strong(value, value | (1 << offset)))
+                    {
+                        return new (GetPoolHead() + (word * kAtomicSize + offset)) T(std::forward<Args>(args)...);
+                    }
+                    else
+                    {
+                        value = usage.load(); // if there is a race, update new usage
+                    }
+                }
+            }
+        }
+        return nullptr;
+    }
+
+    void Delete(T * obj)
+    {
+        size_t at     = static_cast<size_t>(obj - GetPoolHead());
+        size_t word   = at / kAtomicSize;
+        size_t offset = at - (word * kAtomicSize);
+
+#if !defined(NDEBUG)
+        // ensure the obj is in the pool
+        assert(at >= 0);
+        assert(at < N);
+#endif
+
+        obj->~T();
+
+        uint32_t value = mUsage[word].fetch_and(~(1 << offset));
+
+#if !defined(NDEBUG)
+        assert((value & (1 << offset)) != 0); // assert fail when free an unused slot
+#endif
+    }
+
+#if !defined(NDEBUG)
+    template <typename F>
+    void ForEachActiveObject(F f)
+    {
+        for (size_t word = 0; word * kAtomicSize < N; ++word)
+        {
+            auto & usage   = mUsage[word];
+            uint32_t value = usage.load();
+            for (size_t offset = 0; offset < kAtomicSize && offset + word * kAtomicSize < N; ++offset)
+            {
+                if ((value & (1 << offset)) != 0)
+                {
+                    f(GetPoolHead() + (word * kAtomicSize + offset));
+                }
+            }
+        }
+    }
+
+    size_t GetNumObjectsInUse()
+    {
+        size_t count = 0;
+        ForEachActiveObject([&count](T *) { ++count; });
```

Comment:
What will be the type passed to ForEachActiveObject? Is it guaranteed to not use heap or do we not care because debug?

Reply:
#### C++ lambda do not use heap

Lambda is a syntax sugar to generate a anonymous class object. It is essentially the same as following code

```
struct Anonymous
{
    Anonymous(size_t & count) : count(count) {}
    void operator()(T *) { ++count; }
    size_t & count;
};

ForEachActiveObject(Anonymous(count));
```

The function will receive a temporary object of the anonymous class allocated on stack.

#### The function is test-only use

Because there are no thread-safe or atomic way to loop through all bit flags.